File: implementations/python/testing_data_structures.py
"""
Differential and Invariant Testing Framework for Data Structures.

Implements a self-verifying AVL tree candidate and tests it differentially
against Python's built-in set oracle across thousands of randomized mutations,
validating internal balance and binary search invariants at each step.
"""

import random
import unittest
from typing import List, Optional


class AVLNode:
    def __init__(self, key: int):
        self.key = key
        self.height = 1
        self.left: Optional[AVLNode] = None
        self.right: Optional[AVLNode] = None


class AVLTree:
    """Self-verifying AVL tree implementing set semantics."""

    def __init__(self):
        self.root: Optional[AVLNode] = None
        self._size = 0

    def _height(self, node: Optional[AVLNode]) -> int:
        return node.height if node else 0

    def _balance_factor(self, node: Optional[AVLNode]) -> int:
        return self._height(node.left) - self._height(node.right) if node else 0

    def _update_height(self, node: AVLNode) -> None:
        node.height = 1 + max(self._height(node.left), self._height(node.right))

    def _rotate_right(self, y: AVLNode) -> AVLNode:
        x = y.left
        assert x is not None
        t2 = x.right

        x.right = y
        y.left = t2

        self._update_height(y)
        self._update_height(x)
        return x

    def _rotate_left(self, x: AVLNode) -> AVLNode:
        y = x.right
        assert y is not None
        t2 = y.left

        y.left = x
        x.right = t2

        self._update_height(x)
        self._update_height(y)
        return y

    def _balance(self, node: AVLNode) -> AVLNode:
        self._update_height(node)
        bf = self._balance_factor(node)

        # Left heavy
        if bf > 1:
            if self._balance_factor(node.left) < 0:
                assert node.left is not None
                node.left = self._rotate_left(node.left)
            return self._rotate_right(node)

        # Right heavy
        if bf < -1:
            if self._balance_factor(node.right) > 0:
                assert node.right is not None
                node.right = self._rotate_right(node.right)
            return self._rotate_left(node)

        return node

    def insert(self, key: int) -> bool:
        """Inserts a key into the AVL tree. Returns True if inserted, False if duplicate."""
        inserted = False

        def _insert(node: Optional[AVLNode], val: int) -> AVLNode:
            nonlocal inserted
            if not node:
                inserted = True
                return AVLNode(val)

            if val < node.key:
                node.left = _insert(node.left, val)
            elif val > node.key:
                node.right = _insert(node.right, val)
            else:
                inserted = False
                return node

            return self._balance(node)

        self.root = _insert(self.root, key)
        if inserted:
            self._size += 1
        return inserted

    def _min_node(self, node: AVLNode) -> AVLNode:
        curr = node
        while curr.left:
            curr = curr.left
        return curr

    def erase(self, key: int) -> bool:
        """Erases a key from the AVL tree. Returns True if erased, False if not found."""
        erased = False

        def _erase(node: Optional[AVLNode], val: int) -> Optional[AVLNode]:
            nonlocal erased
            if not node:
                erased = False
                return None

            if val < node.key:
                node.left = _erase(node.left, val)
            elif val > node.key:
                node.right = _erase(node.right, val)
            else:
                erased = True
                if not node.left or not node.right:
                    return node.left if node.left else node.right
                else:
                    succ = self._min_node(node.right)
                    node.key = succ.key
                    node.right = _erase(node.right, succ.key)

            return self._balance(node)

        self.root = _erase(self.root, key)
        if erased:
            self._size -= 1
        return erased

    def contains(self, key: int) -> bool:
        curr = self.root
        while curr:
            if key < curr.key:
                curr = curr.left
            elif key > curr.key:
                curr = curr.right
            else:
                return True
        return False

    def size(self) -> int:
        return self._size

    def empty(self) -> bool:
        return self._size == 0

    def to_list(self) -> List[int]:
        out: List[int] = []

        def _in_order(node: Optional[AVLNode]):
            if not node:
                return
            _in_order(node.left)
            out.append(node.key)
            _in_order(node.right)

        _in_order(self.root)
        return out

    def verify_invariants(self) -> bool:
        """Validates strict BST ordering, height metadata, and AVL balance condition."""

        def _verify(node: Optional[AVLNode], low: float, high: float) -> bool:
            if not node:
                return True
            if not (low < node.key < high):
                return False

            lh = self._height(node.left)
            rh = self._height(node.right)

            if node.height != 1 + max(lh, rh):
                return False
            if abs(lh - rh) > 1:
                return False

            return _verify(node.left, low, node.key) and _verify(node.right, node.key, high)

        return _verify(self.root, float("-inf"), float("inf"))


class TestDataStructureVerification(unittest.TestCase):
    def test_boundary_conditions(self):
        tree = AVLTree()
        self.assertTrue(tree.empty())
        self.assertEqual(tree.size(), 0)
        self.assertFalse(tree.contains(10))
        self.assertFalse(tree.erase(10))
        self.assertTrue(tree.verify_invariants())

        # Single element
        self.assertTrue(tree.insert(42))
        self.assertFalse(tree.empty())
        self.assertEqual(tree.size(), 1)
        self.assertTrue(tree.contains(42))
        self.assertTrue(tree.verify_invariants())

        # Duplicate
        self.assertFalse(tree.insert(42))
        self.assertEqual(tree.size(), 1)

        # Removal to empty
        self.assertTrue(tree.erase(42))
        self.assertTrue(tree.empty())
        self.assertEqual(tree.size(), 0)
        self.assertTrue(tree.verify_invariants())

    def test_differential_oracle_stress(self):
        """Differential execution test: candidate AVL vs Python built-in set oracle."""
        rng = random.Random(987654321)
        candidate = AVLTree()
        oracle = set()

        # Execute 5,000 randomized operations
        for step in range(5000):
            op = rng.randint(0, 99)
            key = rng.randint(-300, 300)

            if op < 40:
                # 40% Insert
                cand_res = candidate.insert(key)
                orac_res = key not in oracle
                if orac_res:
                    oracle.add(key)
                self.assertEqual(cand_res, orac_res, f"Insert mismatch at step {step} for key {key}")
            elif op < 70:
                # 30% Erase
                cand_res = candidate.erase(key)
                orac_res = key in oracle
                if orac_res:
                    oracle.remove(key)
                self.assertEqual(cand_res, orac_res, f"Erase mismatch at step {step} for key {key}")
            elif op < 90:
                # 20% Contains
                cand_res = candidate.contains(key)
                orac_res = key in oracle
                self.assertEqual(cand_res, orac_res, f"Contains mismatch at step {step} for key {key}")
            else:
                # 10% Full State / Order comparison
                self.assertEqual(candidate.size(), len(oracle))
                self.assertEqual(candidate.empty(), len(oracle) == 0)
                cand_items = candidate.to_list()
                orac_items = sorted(list(oracle))
                self.assertEqual(cand_items, orac_items)

            # Invariant check
            self.assertTrue(candidate.verify_invariants(), f"Invariant broken at step {step}")

    def test_metamorphic_properties(self):
        """Metamorphic testing: idempotence and round-trip consistency."""
        tree = AVLTree()
        keys = [15, 6, 18, 3, 7, 17, 20, 2, 4, 13, 9]
        for k in keys:
            tree.insert(k)

        orig_list = tree.to_list()
        # Idempotence: re-inserting existing keys produces no state change
        for k in keys:
            self.assertFalse(tree.insert(k))
        self.assertEqual(tree.to_list(), orig_list)

        # Round-trip reversibility: inserting and erasing new key restores state
        tree.insert(999)
        tree.erase(999)
        self.assertEqual(tree.to_list(), orig_list)


if __name__ == "__main__":
    unittest.main()

File: implementations/cpp/testing_data_structures.cpp
/**
 * @file testing_data_structures.cpp
 * @brief Differential testing framework and self-verifying AVL tree candidate vs std::set oracle.
 *
 * Implements an AVL Tree candidate data structure with internal invariant verification,
 * and a differential testing harness executing randomized operations against std::set.
 * Verified with C++17 (-std=c++17 -O3 -Wall -Wextra -Werror).
 */

#include <iostream>
#include <vector>
#include <set>
#include <cstdint>
#include <cassert>
#include <algorithm>
#include <random>

namespace dsa {

/**
 * @brief Self-verifying AVL Tree implementation.
 */
class AVLTree {
private:
    struct Node {
        int key;
        int height;
        Node* left;
        Node* right;

        explicit Node(int val) : key(val), height(1), left(nullptr), right(nullptr) {}
    };

    Node* root_ = nullptr;
    size_t size_ = 0;

    int get_height(Node* n) const {
        return n ? n->height : 0;
    }

    int get_balance(Node* n) const {
        return n ? get_height(n->left) - get_height(n->right) : 0;
    }

    void update_height(Node* n) {
        if (n) {
            n->height = 1 + std::max(get_height(n->left), get_height(n->right));
        }
    }

    Node* rotate_right(Node* y) {
        Node* x = y->left;
        Node* t2 = x->right;

        x->right = y;
        y->left = t2;

        update_height(y);
        update_height(x);

        return x;
    }

    Node* rotate_left(Node* x) {
        Node* y = x->right;
        Node* t2 = y->left;

        y->left = x;
        x->right = t2;

        update_height(x);
        update_height(y);

        return y;
    }

    Node* balance(Node* node) {
        if (!node) return nullptr;
        update_height(node);

        int balance_factor = get_balance(node);

        // Left heavy
        if (balance_factor > 1) {
            if (get_balance(node->left) < 0) {
                node->left = rotate_left(node->left);
            }
            return rotate_right(node);
        }

        // Right heavy
        if (balance_factor < -1) {
            if (get_balance(node->right) > 0) {
                node->right = rotate_right(node->right);
            }
            return rotate_left(node);
        }

        return node;
    }

    Node* insert(Node* node, int key, bool& inserted) {
        if (!node) {
            inserted = true;
            return new Node(key);
        }

        if (key < node->key) {
            node->left = insert(node->left, key, inserted);
        } else if (key > node->key) {
            node->right = insert(node->right, key, inserted);
        } else {
            inserted = false; // Key already present
            return node;
        }

        return balance(node);
    }

    Node* find_min(Node* node) const {
        while (node && node->left) {
            node = node->left;
        }
        return node;
    }

    Node* erase(Node* node, int key, bool& erased) {
        if (!node) {
            erased = false;
            return nullptr;
        }

        if (key < node->key) {
            node->left = erase(node->left, key, erased);
        } else if (key > node->key) {
            node->right = erase(node->right, key, erased);
        } else {
            erased = true;
            if (!node->left || !node->right) {
                Node* temp = node->left ? node->left : node->right;
                if (!temp) {
                    // No child
                    delete node;
                    return nullptr;
                } else {
                    // One child
                    *node = *temp;
                    delete temp;
                }
            } else {
                // Two children
                Node* succ = find_min(node->right);
                node->key = succ->key;
                node->right = erase(node->right, succ->key, erased);
            }
        }

        return balance(node);
    }

    void destroy(Node* node) {
        if (!node) return;
        destroy(node->left);
        destroy(node->right);
        delete node;
    }

    void in_order(Node* node, std::vector<int>& out) const {
        if (!node) return;
        in_order(node->left, out);
        out.push_back(node->key);
        in_order(node->right, out);
    }

    bool verify_node_invariants(Node* node, int64_t min_key, int64_t max_key) const {
        if (!node) return true;

        if (node->key <= min_key || node->key >= max_key) return false;

        int lh = get_height(node->left);
        int rh = get_height(node->right);

        if (node->height != 1 + std::max(lh, rh)) return false;
        if (std::abs(lh - rh) > 1) return false;

        return verify_node_invariants(node->left, min_key, node->key) &&
               verify_node_invariants(node->right, node->key, max_key);
    }

public:
    AVLTree() = default;
    ~AVLTree() {
        destroy(root_);
    }

    // Non-copyable for simplicity
    AVLTree(const AVLTree&) = delete;
    AVLTree& operator=(const AVLTree&) = delete;

    bool insert(int key) {
        bool inserted = false;
        root_ = insert(root_, key, inserted);
        if (inserted) ++size_;
        return inserted;
    }

    bool erase(int key) {
        bool erased = false;
        root_ = erase(root_, key, erased);
        if (erased) --size_;
        return erased;
    }

    bool contains(int key) const {
        Node* curr = root_;
        while (curr) {
            if (key < curr->key) curr = curr->left;
            else if (key > curr->key) curr = curr->right;
            else return true;
        }
        return false;
    }

    size_t size() const {
        return size_;
    }

    bool empty() const {
        return size_ == 0;
    }

    std::vector<int> to_vector() const {
        std::vector<int> out;
        out.reserve(size_);
        in_order(root_, out);
        return out;
    }

    /**
     * @brief Invariant validation method.
     * Verifies strict BST ordering, height tracking, and balance factors across all nodes.
     */
    bool verify_invariants() const {
        return verify_node_invariants(root_, -9223372036854775807LL, 9223372036854775807LL);
    }
};

/**
 * @brief Differential testing engine comparing AVLTree against std::set oracle.
 */
inline void run_differential_tests(uint64_t seed, int operations_count) {
    std::mt19937_64 rng(seed);
    std::uniform_int_distribution<int> op_dist(0, 99);
    std::uniform_int_distribution<int> key_dist(-1000, 1000);

    AVLTree candidate;
    std::set<int> oracle;

    for (int step = 0; step < operations_count; ++step) {
        int op = op_dist(rng);
        int key = key_dist(rng);

        if (op < 40) {
            // 40% Insert
            bool cand_res = candidate.insert(key);
            bool orac_res = oracle.insert(key).second;
            assert(cand_res == orac_res);
        } else if (op < 70) {
            // 30% Erase
            bool cand_res = candidate.erase(key);
            bool orac_res = (oracle.erase(key) > 0);
            assert(cand_res == orac_res);
        } else if (op < 90) {
            // 20% Contains
            bool cand_res = candidate.contains(key);
            bool orac_res = (oracle.find(key) != oracle.end());
            assert(cand_res == orac_res);
        } else {
            // 10% Traversal check
            assert(candidate.size() == oracle.size());
            assert(candidate.empty() == oracle.empty());
            std::vector<int> cand_vec = candidate.to_vector();
            std::vector<int> orac_vec(oracle.begin(), oracle.end());
            assert(cand_vec == orac_vec);
        }

        // Deep internal invariant check after every mutation
        assert(candidate.verify_invariants());
    }
}

} // namespace dsa

int main() {
    std::cout << "Running Data Structure Differential & Invariant Testing..." << std::endl;

    // Boundary / Unit tests
    dsa::AVLTree tree;
    assert(tree.empty());
    assert(tree.size() == 0);
    assert(!tree.contains(42));
    assert(!tree.erase(42));

    // Single element insert / erase
    assert(tree.insert(10));
    assert(!tree.empty());
    assert(tree.size() == 1);
    assert(tree.contains(10));
    assert(tree.verify_invariants());

    assert(!tree.insert(10)); // duplicate
    assert(tree.size() == 1);

    assert(tree.erase(10));
    assert(tree.empty());
    assert(tree.size() == 0);
    assert(tree.verify_invariants());

    // Run differential stress test with 30,000 randomized operations against std::set
    uint64_t seed = 42424242ULL;
    dsa::run_differential_tests(seed, 30000);

    std::cout << "Completed 30,000 differential operations with zero discrepancies!" << std::endl;
    std::cout << "All AVL invariants verified successfully!" << std::endl;
    return 0;
}
